src/main/java/apartment/ui/Controller.java

```diff
@@ -0,0 +1,197 @@
+
+@Component
+public class Controller {
+
+    View view;
+    HostService hostService;
+    ReservationService reservationService;
+    GuestService guestService;
+
+    public Controller(HostService hostService, ReservationService reservationService, GuestService guestService, View view) {
+        this.hostService = hostService;
+        this.reservationService = reservationService;
+        this.guestService = guestService;
+        this.view = view;
+    }
+
+
+    public void run() {
+        view.displayHeader("Welcome to Sustainable Foraging");
+        try {
+            appLoop();
+        } catch (DataException ex) {
+            view.displayException(ex);
+        }
+        view.displayHeader("Goodbye.");
+    }
+
+    public void appLoop() throws DataException {
+        MainMenuOption option;
+        do {
+            option = view.selectMainMenuOption();
+            switch (option) {
+                case VIEW_RESERVATION:
+                    displayViewReservation();
+                    break;
+                case MAKE_RESERVATION:
+                    displayMakeReservation();
+                    break;
+                case EDIT_RESERVATION:
+                    displayEditReservation();
+                    break;
+                case CANCEL_RESERVATION:
+                    displayCancelReservation();
+                    break;
+            }
+        } while (option != MainMenuOption.EXIT);
+    }
+
+    public void displayViewReservation() throws DataException {
+        view.displayHeader(MainMenuOption.VIEW_RESERVATION.getMessage());
+
+        String hostEmail = view.getUserStringInput("Host Email: ");
+
+        Host host = hostService.findHostFromEmail(hostEmail);
+
+        if (host == null) {
+            view.hostNotFound(hostEmail);
+            return;
+        }
+
+        List<Reservation> hostReservations = reservationService.getReservationForParticularHost(host);
+        view.printReservations(host, hostReservations);
+    }
+
+    public void displayCancelReservation() throws DataException {
+        view.displayHeader(MainMenuOption.CANCEL_RESERVATION.getMessage());
+        String guestEmail = view.getUserStringInput("Guest Email: ");
+        Guest guest = guestService.findGuestByEmail(guestEmail);
+        if (guest == null) {
+            view.guestNotFound(guestEmail);
+            return;
+        }
+        String hostEmail = view.getUserStringInput("Host Email: ");
+        Host host = hostService.findHostFromEmail(hostEmail);
+        if (host == null) {
+            view.hostNotFound(hostEmail);
+            return;
+        }
+
+
+        List<Reservation> reservations = reservationService.getReservationForParticularHost(host);
+
+        reservations = reservations.stream().filter(reservation -> reservation.getGuest().getEmail().equals(guestEmail)
+                && reservation.getStartDate().compareTo(LocalDate.now()) > 0).collect(Collectors.toList());
```

Comment:
You could move this method to the service (i.e. `getReservationsForParticularHostAndGuest()`).